### ISSUE TITLE
Drop and re-create view during migrations

### ISSUE THREAD

**Issue by andreievg** [msupply-foundation/open-msupply#4224]
## The suggestion

Quite often when schema changes we have to re-create database views. This is quite a bit of work and can be prone to errors.

@CarlosNZ in Conforma implemented a drop and re-create view in migrations. We can do somethings similar (drop all views before migrations and then re-create them after migration finishes

## Example use case

An example of dropping and re-creating views

### Why should we invest time in this?

Would save time an d have all views in one place with their existing 'state' in code

### Are there any risks associated with this change?

Low risk, most views should be tested already

View tests would need to be done in repo code vs in the migrations, as they won't be available in normal migrations

### How much effort is required?

Up to a day

### Agreed Solution

Drop views at the start of each migration with one script and re-create them at the end of migrations with another script.

**Comment by mark-prins** [comment]
Just curious - how does that handle migrating views? Do you have a way of specifying that a particular view isn't recreated? Is the migration to mutate the 'recreate' function? It's useful to know the full scope of a migration, which includes having the change to the view within it. Perhaps just a comment, to note that the view is changed in this version of the migration?

**Comment by CarlosNZ** [comment]
> Just curious - how does that handle migrating views?

You only need to worry about the current version of the View -- the history of what it used to be is not particularly relevant or helpful (and if it is, just leave a comment)

Do you have a way of specifying that a particular view isn't recreated? 

```sql
DROP VIEW IF EXISTS xyz_view;
-- Deprecated, no longer required
```

> Drop views at the start of each migration with one script and re-create them at the end of migrations with another script.

You can actually go a step further and just have the one script at the end. As the migrations progress, if you're using CASCADE as you drop columns and tables, any connected views automatically get dropped (which can be a problem if you've forgotten about a view linked to a table being modified). So the finishing script just needs to drop all views and re-create the "current" version of them, with explanatory comments.

Might look a little something like this:

```sql
-- View to display combination of various data sources
DROP VIEW IF EXISTS abc;
CREATE VIEW AS abc (
  SELECT a, b, c from some_table...
);

-- This view no longer required
DROP VIEW IF EXISTS xyz;

-- View to do something else
DROP VIEW IF EXISTS my_view;
CREATE VIEW AS my_view (
  SELECT....
)

-- etc...
```

It's quite reassuring to have all the views (and functions if you have them) collected together in one place and know they'll they'll definitely be correct after every migration. And you only have to look in once place to see what they're doing instead of following a complex chain of migration logic.

**Comment by andreievg** [comment]
A couple of notes from refinement

* If the view is completely dropped, we can add the drop to migrations ? (although this may not work since there could be dependencies for that view that change prior ¯\_(ツ)_/¯)
* Would be good if the views (at least the resulting view if it's composed), had a small test, at least to see if the syntax is correct (to make sure the view is still inserted), this can be in repository layer (i think all existing view have this anyways()
* Good to know that this process exists, can just add to migration/README.md

**Comment by roxy-dao** [comment]
Closed in #4517
